jasmine test setup

kashifch opened the pull request:
@benpatterson , @chrisndodge ,

I have created or at least tried to create the jasmine test setup.

I have used gulp-karma and karma-jasmine to make the setup

I have created one simple js file and one test in edx-notifications/static/js folder

@benpatterson , please go through the files whenever you get some time as I am sure there will be some issues.

The setup is working locally but I have to manually install node.js as apt-get commands are not working on my machine, not sure what else will be needed for installations.

In karma.config I have used coverage reporter but I am not sure if it is working properly or not.

Instead of using a make file I have again used a shell script to keep things simpler.

I haven't used the config.js or spec-runner.js

chrisndodge commented:
@kashifch thanks for setting this up. Seems like Travis is failing installing all of the requirements.

kashifch commented:
yes but I though it would fail while installing npm but somehow it installed the npm and the failure is on pip install -r requirements.txt, which is not changed

benpatterson commented on .travis.yml:
```diff
@@ -1,14 +1,19 @@
-language: python
+anguage: python
```
looks like a typo. missing the first l in language.

kashifch replied:
Thanks @benpatterson 

fixing the typo took care of installation stuff

benpatterson commented on .travis.yml:
```diff
@@ -1,14 +1,19 @@
 before_install:
+  - export PATH=./node_modules/.bin:$PATH
+  - sudo apt-get update -qq
+  - sudo apt-get install -y npm
```
I'm curious if this was necessary....we have a private repo running node and we didn't need to install it. Though, the language we're using there is php instead of python.

kashifch replied:
not sure, the analytics-dashboard is using these commands to first install npm so I have included these, not sure if node will be available by default, I guess I can try be commenting these and then see if travis fails or not

benpatterson commented:
Just a couple of comments ... let us know if those fix the issues? I can then take a closer look.

chrisndodge commented:
@kashifch congrats. Seems to work now.

chrisndodge commented on edx_notifications/static/js/HelloWorld.js:
```diff
@@ -0,0 +1,3 @@
+function helloWorld(){
+    return "Hello world";
+}
```
I'd really not to have to add a "HelloWorld.js" file into the codebase, just to verify that we can launch a Jasmine test. Any chance we can get a bare bones test of one aspect of the existing Backbone code in server/web/static/js?

kashifch commented:
@chrisndodge ,

I was just trying to complete the setup, I would definitely change the test, this whole setup is a kind of shot in the dark so I am waiting for @benpatterson review, I think some other changes will be required. For now I will try to comment the npm installation and see if it works without it, I will remove these tests and add an actual one in the real directory tomorrow along with any other changes that will come out in review as necessary

chrisndodge commented on karma.conf.js:
```diff
@@ -0,0 +1,57 @@
+    files: [
+      'edx_notifications/static/js/HelloWorld.js',
+      'edx_notifications/static/js/specs/hello-world-spec.js'
+    ],
```
might there be a way to inspect the files on disk to build up this array of 'files' so that as we build out more Backbone JS files as well as test specs, we don't need to add them here?

Is karma.conf.js a NodeJS executable? If so, it seems like we should be able to access the filesystem.

kashifch replied:
yes we could(and will) use patterns like

 // list of files / patterns to load in the browser
        files: [
            {pattern: 'analytics_dashboard/static/vendor/**/*.js', included: false},
            {pattern: 'analytics_dashboard/static/bower_components/**/*.js', included: false},
            {pattern: 'analytics_dashboard/static/bower_components/**/*.json', included: false},
            {pattern: 'analytics_dashboard/static/js/models/**/*.js', included: false},
            {pattern: 'analytics_dashboard/static/js/views/**/*.js', included: false},
            {pattern: 'analytics_dashboard/static/js/utils/**/*.js', included: false},
            {pattern: 'analytics_dashboard/static/js/test/specs/*.js', included: false},
        ],

chrisndodge commented:
Hi @kashifch just a quick check in, any luck getting a sample test to use existing JS and to use filepatters? Then I think we'll be set to merge this in.

kashifch commented:
Hi @chrisndodge , sorry for the delay but today we faced a lot of errors when using the real files, turned out that the errors were due to dependencies, when patterns are used, jasmine loads some files before which have dependencies on other files and so the scripts just failed, I have tried to find a way around but haven't succeeded yet. I guess this could be solved by the use of requirejs, for now I have included the files in specific order.
@benpatterson, is there any other way to resolve the dependencies issue in short term? also the coverage stats always show 0 lines, 0 functions, am I missing some settings there? 
Overall I would be more comfortable with merging it after a review, made some local changes and still working on a good test, Sorry again for the delay, I guess it is one of those days with too many errors and too little productivity :(

chrisndodge commented:
Thanks @kashifch , I can understand the dependency problem that you descrive. @andy-armstrong has given feedback that we should use RequireJS in our BackboneJS. I'll look into implementing that tonight or tomorrow, but I have to familiarize myself with that tool as I haven't used it personally.

benpatterson commented:
@karadhya In terms of capturing coverage, it looks like adding a 'preprocessors' setting with file paths to the karma.conf.js file should take care of it for you. Doc on that: http://karma-runner.github.io/0.8/config/preprocessors.html

You can also use the dashboard configuration as an example. The preprocessor loads that js in order to trace it while it's being tested. So you'll want the preprocessors to essentially be the same as 'source' for measuring coverage against.

andy-armstrong commented on edx_notifications/server/web/static/edx_notifications/js/test/specs/counter_icon_view_spec.js:
```diff
@@ -0,0 +1,30 @@
+describe("CounterIconView", function(){
```
This will need to be updated to use RequireJS with https://github.com/edx/edx-notifications/pull/16.

andy-armstrong commented on edx_notifications/server/web/static/edx_notifications/js/test/specs/counter_icon_view_spec.js:
```diff
@@ -0,0 +1,30 @@
+    });
+
+    it("counter view model url is set successfully", function(){
+         $('body').append('<div id="edx_notification_container" />');
```
This should use fixtures rather than appending onto the body.

andy-armstrong commented on edx_notifications/server/web/static/edx_notifications/js/test/specs/counter_icon_view_spec.js:
```diff
@@ -0,0 +1,30 @@
+         $('body').append('<div id="edx_notification_container" />');
+        this.counter_view = new CounterIconView({
+            el: $("#edx_notification_container"),
+            pane_el: $("#edx_notification_pane"),
```
It doesn't look like there is a div with "#edx_notification_pane".

andy-armstrong commented on edx_notifications/server/web/static/edx_notifications/js/test/specs/counter_icon_view_spec.js:
```diff
@@ -0,0 +1,30 @@
+
+    it("counter view model url is set successfully", function(){
+         $('body').append('<div id="edx_notification_container" />');
+        this.counter_view = new CounterIconView({
```
I would move the construction of the view into a beforeEach method, so that future tests can share it. It also has the benefit of making the test shorter and easier to understand.

andy-armstrong commented on edx_notifications/server/web/static/edx_notifications/js/vendor/dev/mock-ajax.js:
```diff
@@ -0,0 +1,576 @@
+/*
+
+Jasmine-Ajax : a set of helpers for testing AJAX requests under the Jasmine
```
We are using Sinon in edx-platform for mock AJAX requests. We should probably agree upon a standard library to use in our various repos. Are there advantages of Jasmine-Ajax over Sinon?

kashifch commented:
Hi @andy-armstrong ,

Thanks for the feedback. I really appreciate your help and suggestions as I am completely new to this jasmine thing. I will update the test according to the guidelines you provided.

Thanks,
Kashif

kashifch commented:
Used sinon instead of jasmine-ajax
used fixtures instead of appending div to body
moved the main test to beforeEach section
Added the ability to run tests in debug mode using gulp debug command

The requireJS settings is not implemented, when requireJS is implemented on main repository, we will implement and test.

andy-armstrong commented:
:+1: Looks good to me

chrisndodge commented:
Closing this PR because we are aiming to use RequireJS and that work - including changes for Jasmine support - is at: https://github.com/edx/edx-notifications/pull/17
